### PR TITLE
Detect SVG files by filename extension

### PR DESCRIPTION
**Pull request by arnaud-morvan**
Relate #8 
Relate https://github.com/c2corg/v6_ui/pull/1028

As imagemagick does not seems to correctly returns SVG file format, I use the source filename extension.

### CODE REVIEW

File: c2corg_images/views.py [hunk 1]
```diff
@@ -21,8 +21,13 @@ def ping(request):
     return Response('Pong!' % request.matchdict)
 
 
-def get_format(filename: str) -> str:
-    with Image(filename=filename) as image:
+def get_format(path: str, filename: str) -> str:
+    # imagemajick returns 'PNG' for SVG files
```
Comment (asaunier) [inline c2corg_images/views.py]: imagemagick

**Comment by asaunier** [comment]
How will the image backend behave after this change?
Will it returns 
```
{"filename": "1480094158_257053204.svg"}
```
instead of
```
{"filename": "1480094158_257053204.png"}
```
?
If yes, are the resized images also in SVG?

**Comment by arnaud-morvan** [comment]
Yes, the backend will return ``{"filename": "1480094158_257053204.svg"}``.

But no, the resized images are in ``.jpg``.

**Comment by arnaud-morvan** [comment]
See comments on https://github.com/c2corg/v6_ui/pull/1028#pullrequestreview-10315573

**Comment by asaunier** [comment]
> Yes, the backend will return {"filename": "1480094158_257053204.svg"}.
But no, the resized images are in .jpg.

We have 2 possible approaches (at least) when a new SVG image is uploaded:

1/ the backend returns ``{"filename": "1480094158_257053204.svg"}``
=> no need for ``has_svg`` attribute in the API
=> the UI is responsible to convert the SI/MI/BI filenames to ".jpg" (ie. for most of the occurences of the image in the UI, the original is displayed only when clicking on the BI image in the image detail view).

2/ the backend returns ``{"filename": "1480094158_257053204.jpg", "has_svg": true}``
=> the API stores and passes both data to the UI
=> the UI must convert the original URL to use ".jpg" instead of ".svg"

Approach 1 means no change in the image backend and the API but is more complex on the UI side.
Approach 2 means that the image backend must pass an ``has_svg`` boolean to the API that must it self store it in the database.
Please note that except for the image backend <-> API communication, I have already created PR in the UI and API.

A problem with approach 1 though: in the v5 database filenames of SVG images are stored using the converted JPEG (or even PNG!) formats. See for instance:
```
c2corg=# select filename, has_svg from images where has_svg is true limit 10;
         filename          | has_svg 
---------------------------+---------
 1295521091_782327294.jpg  | t
 1295521186_407896485.jpg  | t
 1300544886_1578303043.jpg | t
 1291623090_634096000.jpg  | t
 1303390400_1718177378.jpg | t
 1274781276_1156353716.jpg | t
 1275116587_704216532.jpg  | t
 1275117169_24903222.jpg   | t
 1275480380_1134138993.png | t
 1277128652_1921461223.png | t
(10 rows)
```
=> those data are transferred as is in the v6 database during the migration: https://github.com/c2corg/v6_api/blob/master/c2corg_api/scripts/migration/documents/images.py#L74
Of course we could change the migration script so that ``filename`` is converted when ``has_svg`` is detected in v5 data.
Please note in the data examples above that some SVG seems to have been converted to PNG (?).  Perhaps it's not a problem because the image migration script converts all SVG to JPEG? Thus those "png" ``filename`` entries must anyway be converted while migrationg the data?

**Comment by arnaud-morvan** [comment]
IMHO, It seems logicial to have the good extension on filenames (corresponding to the file content).

**Comment by arnaud-morvan** [comment]
Does the V5 store original image as SVG and JPG format for the same image (original size) ?

**Comment by asaunier** [comment]
> Does the V5 store original image as SVG and JPG format for the same image (original size) ?

It seems so. For instance http://www.camptocamp.org/images/208579/fr/face-n-des-grandes-jorasses-quelques-itineraires
=> there are both
* http://s.camptocamp.org/uploads/images/1336219429_894641918.jpg
* http://s.camptocamp.org/uploads/images/1336219429_894641918.svg

I think in v6 we only need the SVG original, not the JPEG.
How is this case managed by the images migration script?

**Comment by arnaud-morvan** [comment]
I was not aware of this, I will update the migration script to take care of svg files, based on has_svg field.

> I  think in v6 we only need the SVG original, not the JPEG.

Yes, but it seems simpler for the UI part to handle "jpg" filename comin from the API.
And it seems little awkward to me to return a filename that do not really exists.

**Comment by asaunier** [comment]
> Yes, but it seems simpler for the UI part to handle "jpg" filename comin from the API.

I don't understand what you mean.
The UI does not need the JPEG "original" because it can serves the SVG original instead.
We use JPEG resized images because it seems not that easy to resize or crop SVG images (with an embedded JPEG image), is it?

> And it seems little awkward to me to return a filename that do not really exists.

Right. Actually the more consistent way would perhaps have been to return the filename (with no extension) and the extension in dedicated attributes (which would have helped building the URLs of the resized images).

If we can assume that all SVG originals will have JPEG resized versions (including for the migrated-from-v5 images), we can indeed use approach 1. I will then cancel the PRs I have recently proposed and make changes:
* to the image migration script in the API
* to have a tool used to generate the image URLs (for all formats) in the UI.

**Comment by asaunier** [comment]
> And it seems little awkward to me to return a filename that do not really exists.

I am a bit concerned that your approach implies quite a lot of changes both in the UI and the API:
* in the detail view of images to make the BI image use the JPEG while it links to the SVG (OK, no big deal we need to do changes here anyway)
* in the image cards (+ in feed items with images), in order to change the ``.svg`` extension to ``.jpg`` in https://github.com/c2corg/v6_ui/blob/master/c2corg_ui/static/js/card.js#L147 and https://github.com/c2corg/v6_ui/blob/master/c2corg_ui/static/js/utils.js#L158
* in the UI carousel and slideshow in order to change the ``.svg`` extension to ``.jpg``. @ginold could you help me and point were the image filenames are handled in the carousel and slideshow?
* in the API proxy https://github.com/c2corg/v6_api/blob/592903ddae28c3b1edeac5cbdabff185027ffe46/c2corg_api/views/image.py#L206-L226 => a change is needed anyway so that the API proxy returns the SVG when requested for the original, else JPEG

=> a lot of changes required whereas we generally manipulate resized images (ie. JPEG) instead of SVG. Actually we need the SVG filename only in the image detail view when linking towards the original file (SVG).

**Comment by arnaud-morvan** [comment]
So the simpler way should be to return the jpg extension with has_svg as in V5 (but that file won't exist in the S3 bucket, no need to convert the original image to jpg as it will never be used.

We only have to : 
* Change the extension returned by the backend
* Rectify the migration script to take care of has_svg.

**Comment by ginold** [comment]
>  could you help me and point were the image filenames are handled in the carousel and slideshow?

The only thing that is happening is the function `app.utils.createImageUrl` that adds SI/MI/BI before the ending `.extension`, there is no function yet that converts the extension itself.

1) if you load the view page:
`loadImages_` is called and will load the associated and embedded images, create slides with `app.utils.createImageSlide` which calls inside `app.utils.createImageUrl`.

2) if you upload an image in the view page:
after you click on save, `app.ImageUploaderController.prototype.save` is called and will create slides with `app.utils.createImageSlide` and compile them (similar to the above)

**Comment by asaunier** [comment]
> We only have to :
Change the extension returned by the backend
Rectify the migration script to take care of has_svg 

Actually your original approach requires far less work then I expected.

According to @ginold the UI JS changes would be limited to adapt https://github.com/c2corg/v6_ui/blob/master/c2corg_ui/static/js/utils.js#L158-L163

Other changes in the UI are not that big.

Concerning the change during the migration I had already prepared something yesterday.

A change in the API image proxy is needed anyway, whatever the approach (I will do it).

So it's OK to keep your original approach and simply make the image backend return the SVG filename.
Let me know what you prefer.

**Comment by asaunier** [comment]
> So it's OK to keep your original approach and simply make the image backend return the SVG filename.

Actually I think this approach is more consistent because the original image IS an SVG, so it's weird to store the filename as "jpg" in the API, and finally showing jpeg resized images is somehow a kind of workaround => perhaps in the future we will use SVG for all sizes.

In addition it seems that it was not that a big deal to make the changes in the UI and API, see
https://github.com/c2corg/v6_api/pull/548
https://github.com/c2corg/v6_ui/pull/1028

**Pull request merged**